### ISSUE TITLE
CI: MinimumMasterNodesIT testMultipleNodesShutdownNonMasterNodes fails

### ISSUE THREAD

**Issue by jaymode** [elastic/elasticsearch#23922]
The testMultipleNodesShutdownNonMasterNodes test has failed intermittently in CI on master and 5.x. Looking back I see two types of failures. The first is:

```
ERROR   6.18s J2 | MinimumMasterNodesIT.testMultipleNodesShutdownNonMasterNodes <<< FAILURES!
   > Throwable #1: ClusterBlockException[blocked by: [SERVICE_UNAVAILABLE/1/state not recovered / initialized];]
   > 	at __randomizedtesting.SeedInfo.seed([D9FA00AB1872EEF1:921B6790A81064E5]:0)
   > 	at org.elasticsearch.cluster.block.ClusterBlocks.globalBlockedException(ClusterBlocks.java:165)
   > 	at org.elasticsearch.cluster.block.ClusterBlocks.globalBlockedRaiseException(ClusterBlocks.java:151)
   > 	at org.elasticsearch.action.search.TransportSearchAction.executeSearch(TransportSearchAction.java:210)
   > 	at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:190)
   > 	at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:59)
   > 	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:170)
   > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:142)
   > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:84)
   > 	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:83)
   > 	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:72)
   > 	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:408)
   > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:80)
   > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:54)
   > 	at org.elasticsearch.cluster.MinimumMasterNodesIT.testMultipleNodesShutdownNonMasterNodes(MinimumMasterNodesIT.java:265)
   > 	at java.lang.Thread.run(Thread.java:745)
```

Reproduce lines but neither reproduced for me:

```
gradle :core:integTest -Dtests.seed=D9FA00AB1872EEF1 -Dtests.class=org.elasticsearch.cluster.MinimumMasterNodesIT -Dtests.method="testMultipleNodesShutdownNonMasterNodes" -Dtests.security.manager=true -Dtests.locale=tr-TR -Dtests.timezone=Pacific/Majuro
```

```
gradle :core:integTest -Dtests.seed=E771366E0D357966 -Dtests.class=org.elasticsearch.cluster.MinimumMasterNodesIT -Dtests.method="testMultipleNodesShutdownNonMasterNodes" -Dtests.security.manager=true -Dtests.locale=sl -Dtests.timezone=America/Chicago
```

Failures:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-unix-compatibility/os=amazon/655/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+periodic/2287/consoleText

The second failure is:

```
FAILURE 7.36s J0 | MinimumMasterNodesIT.testMultipleNodesShutdownNonMasterNodes <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Count is 68 but 100 was expected.  Total shards: 3 Successful shards: 2 & 1 shard failures:
   >  shard [[GzNESONVQ8GhAsKH164iDw][test][0]], reason [NodeNotConnectedException[[node_t3][127.0.0.1:9443] Node not connected]], cause [NodeNotConnectedException[[node_t3][127.0.0.1:9443] Node not connected]
   > 	at org.elasticsearch.transport.TcpTransport.getConnection(TcpTransport.java:602)
   > 	at org.elasticsearch.transport.TcpTransport.getConnection(TcpTransport.java:116)
   > 	at org.elasticsearch.test.transport.MockTransportService$LookupTestTransport.getConnection(MockTransportService.java:521)
   > 	at org.elasticsearch.transport.TransportService.getConnection(TransportService.java:509)
   > 	at org.elasticsearch.action.search.SearchTransportService.getConnection(SearchTransportService.java:402)
   > 	at org.elasticsearch.action.search.TransportSearchAction.lambda$executeSearch$3(TransportSearchAction.java:257)
   > 	at org.elasticsearch.action.search.AbstractSearchAsyncAction.getConnection(AbstractSearchAsyncAction.java:276)
   > 	at org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.executePhaseOnShard(SearchQueryThenFetchAsyncAction.java:77)
   > 	at org.elasticsearch.action.search.InitialSearchPhase.performPhaseOnShard(InitialSearchPhase.java:146)
   > 	at org.elasticsearch.action.search.InitialSearchPhase.run(InitialSearchPhase.java:131)
   > 	at org.elasticsearch.action.search.AbstractSearchAsyncAction.executePhase(AbstractSearchAsyncAction.java:144)
   > 	at org.elasticsearch.action.search.AbstractSearchAsyncAction.start(AbstractSearchAsyncAction.java:114)
   > 	at org.elasticsearch.action.search.TransportSearchAction.executeSearch(TransportSearchAction.java:268)
   > 	at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:190)
   > 	at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:59)
   > 	at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:167)
   > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:139)
   > 	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:81)
   > 	at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:83)
   > 	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:72)
   > 	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:405)
   > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:77)
   > 	at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:51)
   > 	at org.elasticsearch.cluster.MinimumMasterNodesIT.testMultipleNodesShutdownNonMasterNodes(MinimumMasterNodesIT.java:262)
   > 	at java.lang.Thread.run(Thread.java:745)
   > ]
   > 	at __randomizedtesting.SeedInfo.seed([601D0F659548D429:2BFC685E252A5E3D]:0)
   > 	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertHitCount(ElasticsearchAssertions.java:257)
   > 	at org.elasticsearch.cluster.MinimumMasterNodesIT.testMultipleNodesShutdownNonMasterNodes(MinimumMasterNodesIT.java:262)
   > 	at java.lang.Thread.run(Thread.java:745)
```

This one did not reproduce either but the line from the failure is:

```
gradle :core:integTest -Dtests.seed=E771366E0D357966 -Dtests.class=org.elasticsearch.cluster.MinimumMasterNodesIT -Dtests.method="testMultipleNodesShutdownNonMasterNodes" -Dtests.security.manager=true -Dtests.locale=sl -Dtests.timezone=America/Chicago
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+periodic/2296/consoleText

**Comment by ywelsch** [comment]
The first failure happens for the following reason:
- 4 nodes (nodes 0, 1, 2, and 3) running with `minimum_master_nodes` set to 3
- we stop 2 nodes (node 0 and 3)
- wait for cluster block to be in place on all nodes
- start 2 nodes (node 4 and node 5) and do a `prepareHealth().setWaitForNodes("4")`
- then do a search request

The search request runs into the `ClusterBlockException` as the `prepareHealth().setWaitForNodes("4")` check succeeds on a cluster state that has nodes 1, 2, 3, and 4, i.e., only one of the two new nodes has joined the cluster and only one of the two dead nodes was removed by the master (removing the dead nodes only happens after there are again `minimum_master_nodes` nodes in the cluster).

I pushed a3cceb8a00 which uses a method from `InternalTestCluster` that checks that the right nodes have rejoined the cluster.

The second failure is similar in nature, here the search fails for a shard on a dead node that has not been booted yet off the cluster. The fix above also fixes that situation.

**Comment by javanna** [comment]
@ywelsch is this recent failure related or something completely different? https://elasticsearch- ci.elastic.co/job/elastic+elasticsearch+5.x+multijob-darwin-compatibility/441

**Comment by ywelsch** [comment]
@javanna that failure is different:

```
Caused by: java.lang.AssertionError: setting value but not in fetching mode
	at __randomizedtesting.SeedInfo.seed([362DF4A99D9305E5]:0)
	at org.elasticsearch.gateway.AsyncShardFetch$NodeEntry.doneFetching(AsyncShardFetch.java:372)
	at org.elasticsearch.gateway.AsyncShardFetch.processAsyncFetch(AsyncShardFetch.java:206)
	at org.elasticsearch.gateway.AsyncShardFetch$1.onResponse(AsyncShardFetch.java:275)
	at org.elasticsearch.gateway.AsyncShardFetch$1.onResponse(AsyncShardFetch.java:272)
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:88)
	at org.elasticsearch.action.support.TransportAction$1.onResponse(TransportAction.java:84)
```

It's a duplicate of #24007. I'll talk to @dakrone to see if we can get this fixed today.

**Comment by ywelsch** [comment]
@javanna it's fixed here: #24434
